File: python/protein-translation/protein_translation.py
# map protein to list of codons
P2C = {
    'Methionine': ['AUG'],
    'Phenylalanine': ['UUU', 'UUC'],
    'Leucine': ['UUA', 'UUG'],
    'Serine': ['UCU', 'UCC', 'UCA', 'UCG'],
    'Tyrosine': ['UAU', 'UAC'],
    'Cysteine': ['UGU', 'UGC'],
    'Tryptophan': ['UGG'],
    'STOP': ['UAA', 'UAG', 'UGA'],
}

# map codon to protein
C2P = {c: p for p in P2C for c in P2C[p]}


def proteins(strand):
    proteins = []
    for i in range(0, len(strand), 3):
        codon = strand[i:i+3]
        protein = C2P[codon]
        if protein == 'STOP':
            break
        proteins.append(protein)
    return proteins

File: php/protein-translation/ProteinTranslation.php
<?php

declare(strict_types=1);

class ProteinTranslation
{
    private const STOP = 'STOP';

    public function getProteins(string $input): array
    {
        $codons = str_split($input, 3);
        $proteins = [];
        foreach ($codons as $codon) {
            $protein = $this->getProtein($codon);
            if ($protein == self::STOP) {
                break;
            }
            $proteins[] = $protein;
        }
        return $proteins;
    }

    public function getProtein(string $codon): string
    {
        return match ($codon) {
            'AUG'                      => 'Methionine',
            'UUU', 'UUC'               => 'Phenylalanine',
            'UUA', 'UUG'               => 'Leucine',
            'UCU', 'UCC', 'UCA', 'UCG' => 'Serine',
            'UAU', 'UAC'               => 'Tyrosine',
            'UGU', 'UGC'               => 'Cysteine',
            'UGG'                      => 'Tryptophan',
            'UAA', 'UAG', 'UGA'        => self::STOP,
            default => throw new InvalidArgumentException('Invalid codon'),
        };
    }
}
